### PR TITLE
Partition LRU key history to reduce contention.

### PR DESCRIPTION
**Pull request by Alan-Jowett**
## Description

This pull request introduces significant changes to the `ebpf_maps.c` file, primarily focusing on enhancing the Least Recently Used (LRU) map's performance and memory usage. The key changes include the introduction of LRU partitions, changes in how key states are managed, and modifications to the LRU entry initialization and uninitialization.

Key changes:

1. Introduction of LRU Partitions:
   * A new constant `EBPF_LRU_MAXIMUM_PARTITIONS` has been added to limit the maximum number of partitions to 8 as a trade-off between memory usage and performance. Fewer partitions will result in more contention on the lock, but more partitions will consume more memory.
   * The `_ebpf_core_lru_map` structure has been modified to include an array of LRU partitions. Each partition has its own hot and cold lists, lock, current generation, and hot list size and limit. [[1]](diffhunk://#diff-ae8a7da4edcb9e1f7c73d5704cb00d475c4f847515c8d97c08a178a326ccb27cR46-R162) [[2]](diffhunk://#diff-ae8a7da4edcb9e1f7c73d5704cb00d475c4f847515c8d97c08a178a326ccb27cR1250)
   * The `_create_lru_hash_map` function now calculates the number of partitions based on the CPU count and initializes each partition separately. [[1]](diffhunk://#diff-ae8a7da4edcb9e1f7c73d5704cb00d475c4f847515c8d97c08a178a326ccb27cR1250) [[2]](diffhunk://#diff-ae8a7da4edcb9e1f7c73d5704cb00d475c4f847515c8d97c08a178a326ccb27cL1153-R1266) [[3]](diffhunk://#diff-ae8a7da4edcb9e1f7c73d5704cb00d475c4f847515c8d97c08a178a326ccb27cL1179-R1314)

2. Changes in Key State Management:
   * The `_get_key_state` function now takes an additional `partition` parameter and checks the key state for the specific partition.
   * The `_merge_hot_into_cold_list_if_needed` and `_insert_into_hot_list` functions have been modified to operate on a specific partition.

3. Changes in LRU Entry Initialization and Uninitialization:
   * The `_initialize_lru_entry` function now initializes an LRU entry for a specific partition and inserts it into the partition's hot list.
   * The `_uninitialize_lru_entry` function has been modified to uninitialize an LRU entry from all partitions.
   * The `_lru_hash_table_notification` function now maps the current CPU to a partition and performs operations based on the notification type for that partition.

4. Additional Changes:
   * A new function `_reap_lru_cold_lists` has been added to walk the cold lists, remove secondary keys, and find the oldest primary key.
   * The `_create_lru_hash_map` function now calculates the size of the LRU map based on the number of partitions. [[1]](diffhunk://#diff-ae8a7da4edcb9e1f7c73d5704cb00d475c4f847515c8d97c08a178a326ccb27cL1153-R1266) [[2]](diffhunk://#diff-ae8a7da4edcb9e1f7c73d5704cb00d475c4f847515c8d97c08a178a326ccb27cR1280-R1292)

## Testing

CI/CD + Performance testing.

```
Baseline:
Timestamp,Test,Average Duration (ns),CPU 0 Duration (ns),CPU 1 Duration (ns),CPU 2 Duration (ns),CPU 3 Duration (ns),CPU 4 Duration (ns),CPU 5 Duration (ns)
2024-04-16T16:10:05-0800,BPF_MAP_TYPE_LRU_HASH read,1944,1953,1959,1938,1948,1915,1953
2024-04-16T16:10:24-0800,BPF_MAP_TYPE_LRU_HASH update,3613,3493,3620,3660,3634,3655,3616
2024-04-16T16:11:01-0800,BPF_MAP_TYPE_LRU_HASH update and read,2000,2327,1942,1944,1916,1940,1934
2024-04-16T16:11:24-0800,BPF_MAP_TYPE_LRU_HASH replace,3652,3521,3664,3707,3690,3690,3642
2024-04-16T16:12:02-0800,BPF_MAP_TYPE_LRU_HASH rolling update,326,327,322,325,326,328,328

New:
Timestamp,Test,Average Duration (ns),CPU 0 Duration (ns),CPU 1 Duration (ns),CPU 2 Duration (ns),CPU 3 Duration (ns),CPU 4 Duration (ns),CPU 5 Duration (ns)
2024-04-17T00:14:21-0800,BPF_MAP_TYPE_LRU_HASH read,212,241,204,202,200,203,224
2024-04-17T00:14:23-0800,BPF_MAP_TYPE_LRU_HASH update,2683,2607,2696,2707,2717,2690,2685
2024-04-17T00:14:51-0800,BPF_MAP_TYPE_LRU_HASH update and read,320,1153,151,145,152,150,169
2024-04-17T00:15:02-0800,BPF_MAP_TYPE_LRU_HASH replace,3068,2985,3083,3088,3096,3082,3075
2024-04-17T00:15:34-0800,BPF_MAP_TYPE_LRU_HASH rolling update,168,160,170,161,171,167,181
```

## Documentation

Documentation in the code.

## Installation

No.

### CODE REVIEW

File: libs/execution_context/ebpf_maps.c [hunk 1]
```diff
@@ -43,33 +43,123 @@ typedef struct _ebpf_core_object_map
...
     size_t hot_list_limit;     //< Maximum size of the hot list.
+} ebpf_lru_partition_t;
+
+/**
+ * @brief The map definition for an LRU map.
+ */
+typedef struct _ebpf_core_lru_map
+{
+    ebpf_core_map_t core_map;           //< Core map structure.
+    size_t partition_count;             //< Number of LRU partitions. Currently a single partition is supported.
```
Comment (mtfriesen) [inline libs/execution_context/ebpf_maps.c:160]: Is this comment still accurate (and similar comment two lines below)?

> `#define EBPF_LRU_MAXIMUM_PARTITIONS 8`

**Review by mtfriesen — DISMISSED** [review]

**Review by mtfriesen — APPROVED** [review]

File: libs/execution_context/ebpf_maps.c [hunk 2]
```diff
@@ -1067,20 +1165,26 @@ _insert_into_hot_list(_Inout_ ebpf_core_lru_map_t* map, _Inout_ ebpf_lru_entry_t
  * @param[in] key Key to initialize the entry with.
  */
 static void
-_initialize_lru_entry(_Inout_ ebpf_core_lru_map_t* map, _Inout_ ebpf_lru_entry_t* entry, _In_ const uint8_t* key)
+_initialize_lru_entry(
+    _Inout_ ebpf_core_lru_map_t* map, _Inout_ ebpf_lru_entry_t* entry, size_t partition, _In_ const uint8_t* key)
 {
-    ebpf_lock_state_t state = ebpf_lock_lock(&map->lock);
-    ebpf_assert(_get_key_state(map, entry) == EBPF_LRU_KEY_UNINITIALIZED);
+    memcpy(EBPF_LRU_ENTRY_KEY_PTR(map, entry), key, map->core_map.ebpf_map_definition.key_size);
```
Comment (shpalani) [inline libs/execution_context/ebpf_maps.c:1171]: For my understanding, can you help clarify the purpose of doing memcpy here?
And also, I think lock is needed during initialization, to protect this critical section.

File: libs/execution_context/ebpf_maps.c [hunk 3]
```diff
@@ -1140,6 +1249,7 @@ _create_lru_hash_map(
 {
     ebpf_result_t retval = EBPF_SUCCESS;
     ebpf_core_lru_map_t* lru_map = NULL;
+    uint32_t partition_count = min(ebpf_get_cpu_count(), EBPF_LRU_MAXIMUM_PARTITIONS);
```
Comment (shpalani) [inline libs/execution_context/ebpf_maps.c:1252]: okay, here the partition count gets the min between CPU cores and max partitions 8.

I have seen in VMs with 40 cores (vcpu). Will there be still contention? 

Use-case: In Server Container Networking.

File: libs/execution_context/ebpf_maps.c [hunk 4]
```diff
@@ -1209,6 +1340,74 @@ _delete_lru_hash_map(_In_ _Post_invalid_ ebpf_core_map_t* map)
...
+ */
+static ebpf_lru_entry_t*
+_reap_lru_cold_lists(ebpf_core_lru_map_t* lru_map)
+{
+    ebpf_lru_entry_t* oldest_entry = NULL;
+    uint64_t oldest_timestamp = UINT64_MAX;
+    bool non_empty_partition_found = true;
+
+    // Scan all partitions until we find a non-empty cold list or we have an oldest entry.
+    while (!oldest_entry && non_empty_partition_found) {
```
Comment (shpalani) [inline libs/execution_context/ebpf_maps.c:1357]: Isn't the whole block a critical section to lock, starting from scan and update?

File: libs/execution_context/ebpf_maps.c [hunk 5]
```diff
@@ -43,33 +43,124 @@ typedef struct _ebpf_core_object_map
 // Define the granularity at which the LRU list is updated.
 #define EBPF_LRU_GENERATION_COUNT 10
 
+// Limit maximum number of partitions to 8 as a trade-off between memory usage and performance.
```
[the remark this answers is not in the document]
Comment (Alan-Jowett) [reply libs/execution_context/ebpf_maps.c:46]: Primarily empirical.

Several reasons:
1) 8 entries of 64bit == 1 cache line.
2) The time to reap entries is a O(N*N) where N == number of partitions. 
3) Partition = 1 is basically the old behavior.

File: libs/execution_context/ebpf_maps.c (hunk 2 above)
Comment (Alan-Jowett) [reply libs/execution_context/ebpf_maps.c:1171]: Each value stored in the hash-table has a supplemental value associated with it. These supplemental values are used to build the key-history lists. Given a supplemental value, there is no easy way to determine what the original key is, so we copy the key value into the supplemental value field.

File: libs/execution_context/ebpf_maps.c (hunk 3 above)
Comment (Alan-Jowett) [reply libs/execution_context/ebpf_maps.c:1252]: This is a tradeoff between memory footprint and performance. If we have data that this is an issue we can try raising it.

File: libs/execution_context/ebpf_maps.c (hunk 4 above)
Comment (Alan-Jowett) [reply libs/execution_context/ebpf_maps.c:1357]: No, CPUs are free to execute this concurrently. This may result in multiple threads attempting to evict the same key, but this doesn't really justify the cost of adding a map wide lock here.

File: libs/execution_context/ebpf_maps.c [hunk 6]
```diff
@@ -43,33 +43,124 @@ typedef struct _ebpf_core_object_map
...
+ * @brief The BPF_MAP_TYPE_LRU_HASH is a hash table that stores a limited number of entries. When the map is full, the
+ * least recently used entry is removed to make room for a new entry. The map is implemented as a hash table with a pair
+ * of double-linked lists per partition. The hot list contains entries that have been accessed in the current
+ * generation, and the cold list contains entries that have not been accessed in the current generation. When entries in
+ * the cold list are accessed, they are moved to the hot list. When the hot list reaches
  * max_entries/EBPF_LRU_GENERATION_COUNT, the hot list is merged into the cold list, a new generation is started, and
- * the hot list is cleared.  When space is needed in the map, the cold list is trimmed to make room for new entries. The
- * cold list is trimmed by removing the oldest entries in the list, which are part of the oldest generation. The benefit
- * is that the lock only needs to be acquired when moving items between generations, and not on every access.
+ * the hot list is cleared. When space is needed an entry is selected from a cold-list and is removed from the
```
Comment (dthaler) [inline libs/execution_context/ebpf_maps.c:57]:
```suggestion
 * the hot list is cleared. When space is needed an entry is selected from a cold list and is removed from the
```
Consistency with lines 54-56

File: libs/execution_context/ebpf_maps.c [hunk 7]
```diff
@@ -43,33 +43,124 @@ typedef struct _ebpf_core_object_map
...
+ * hash-table.
```
Comment (dthaler) [inline libs/execution_context/ebpf_maps.c:58]:
```suggestion
 * hash table.
```
Consistency with line 51 etc.

File: libs/execution_context/ebpf_maps.c [hunk 8]
```diff
@@ -43,33 +43,124 @@ typedef struct _ebpf_core_object_map
...
+ *
+ * Key-history is stored along with the value in the map. The hash-table then provides callbacks to the map to update
```
Comment (dthaler) [inline libs/execution_context/ebpf_maps.c:60]:
```suggestion
 * Key-history is stored along with the value in the map. The hash table then provides callbacks to the map to update
```

File: libs/execution_context/ebpf_maps.c [hunk 9]
```diff
@@ -43,33 +43,124 @@ typedef struct _ebpf_core_object_map
...
+ * the key-history when an entry is accessed, updated, or deleted.
+ *
+ * Key-history can be in multiple partitions, with different generation and last-used-time values. To determine
+ * the actual last used time of a key, the map must iterate over all the partitions and find the maximum last-used-time.
+ *
+ * The link in the ebpf_lru_entry_t with the highest corresponding last-used-time is the actual last used time of the
+ * key. All other links in that ebpf_lru_entry_t are considered stale and are ignored.
+ *
+ * Key history is stored in a separate partition from the value as is returned as an opaque blob of memory.
```
Comment (dthaler) [inline libs/execution_context/ebpf_maps.c:69]: This sentence is hard to parse (around "as is"), can you reword?

File: libs/execution_context/ebpf_maps.c [hunk 10]
```diff
@@ -43,33 +43,124 @@ typedef struct _ebpf_core_object_map
...
+ *
+ * The implementation then treats the opaque blob as a block of memory containing the following:
+ * ebpf_lru_entry_partition_t is an untyped block of memory containing the following:
```
Comment (dthaler) [inline libs/execution_context/ebpf_maps.c:72]: 72 and 71 seem to be redundant

File: libs/execution_context/ebpf_maps.c [hunk 11]
```diff
@@ -43,33 +43,124 @@ typedef struct _ebpf_core_object_map
...
+ * the cold list are accessed, they are moved to the hot list. When the hot list reaches
  * max_entries/EBPF_LRU_GENERATION_COUNT, the hot list is merged into the cold list, a new generation is started, and
- * the hot list is cleared.  When space is needed in the map, the cold list is trimmed to make room for new entries. The
- * cold list is trimmed by removing the oldest entries in the list, which are part of the oldest generation. The benefit
- * is that the lock only needs to be acquired when moving items between generations, and not on every access.
+ * the hot list is cleared. When space is needed an entry is selected from a cold-list and is removed from the
+ * hash-table.
+ *
+ * Key-history is stored along with the value in the map. The hash-table then provides callbacks to the map to update
+ * the key-history when an entry is accessed, updated, or deleted.
```
Comment (dthaler) [inline libs/execution_context/ebpf_maps.c:61]: "key-history" is hyphenated in this paragraph but not in 86, 91, etc.  Pick one and be consistent.

File: libs/execution_context/ebpf_maps.c [hunk 12]
```diff
@@ -1150,8 +1260,12 @@ _create_lru_hash_map(
         goto Exit;
     }
 
-    size_t lru_entry_size;
-    retval = ebpf_safe_size_t_add(EBPF_OFFSET_OF(ebpf_lru_entry_t, key), map_definition->key_size, &lru_entry_size);
+    // Base size of an LRU entry is the size of list entry pointers plus the last used time plus the generation for each
+    // partition.
+    size_t lru_entry_size = (sizeof(ebpf_list_entry_t) + sizeof(size_t) + sizeof(size_t)) * partition_count;
```
Comment (dthaler) [inline libs/execution_context/ebpf_maps.c:1265]: Why can't you use EBPF_LRU_ENTRY_SIZE here?

File: libs/execution_context/ebpf_maps.c [hunk 13]
```diff
@@ -43,33 +43,125 @@ typedef struct _ebpf_core_object_map
...
+ *
+ * ebpf_lru_entry_partition_t is an untyped block of memory containing the following:
+ * struct ebpf_lru_entry_t {
+ *  ebpf_list_entry_t list_entry[partition_count];
+ *  size_t generation[partition_count];
+ *  size_t last_used_time[partition_count];
+ *  uint8_t key[key_size];
+ * };
+ *
+ * This structure is not expressible in C, so we use macros to calculate the offsets of the fields in the structure and
```
Comment (dthaler) [inline libs/execution_context/ebpf_maps.c:79]: Why not group all fields in the same partition together, as in:

```
struct ebpf_lru_entry_partition_t {
    ebpf_list_entry_t list_entry;
    size_t generation;
    size_t last_used_time;
    uint8_t key;
};
struct ebpf_lru_entry_t {
    struct ebpf_lru_entry_partition_t partition[partition_count];
};
```

Comment (Alan-Jowett) [reply libs/execution_context/ebpf_maps.c:79]: Two issues:
1) We can't have 2 variable sized fields in a C struct (both the partitions and the key are variable sized).
2) Computing the effective timestamp for a key is run once per partition, so I chose to make sure the last_used_time are all on the same cache line.

**Review by mikeagun — APPROVED** [review]

**Review by dthaler — APPROVED** [review]

File: libs/execution_context/ebpf_maps.c [hunk 14]
```diff
@@ -1094,21 +1199,24 @@ _initialize_lru_entry(_Inout_ ebpf_core_lru_map_t* map, _Inout_ ebpf_lru_entry_t
...
 _uninitialize_lru_entry(_Inout_ ebpf_core_lru_map_t* map, _Inout_ ebpf_lru_entry_t* entry)
 {
-    ebpf_lock_state_t state = ebpf_lock_lock(&map->lock);
-    ebpf_lru_key_state_t key_state = _get_key_state(map, entry);
-    ebpf_assert(key_state == EBPF_LRU_KEY_HOT || key_state == EBPF_LRU_KEY_COLD);
+    for (size_t partition = 0; partition < map->partition_count; partition++) {
+        ebpf_lock_state_t state = ebpf_lock_lock(&map->partitions[partition].lock);
+        ebpf_lru_key_state_t key_state = _get_key_state(map, partition, entry);
+        ebpf_assert(
+            key_state == EBPF_LRU_KEY_HOT || key_state == EBPF_LRU_KEY_COLD || key_state == EBPF_LRU_KEY_UNINITIALIZED);
```
Comment (mikeagun) [inline libs/execution_context/ebpf_maps.c:1206]: could valid code uninitialize an UNINITIALIZED entry?

File: libs/execution_context/ebpf_maps.c [hunk 15]
```diff
@@ -1209,6 +1337,74 @@ _delete_lru_hash_map(_In_ _Post_invalid_ ebpf_core_map_t* map)
...
+                ebpf_lock_unlock(&lru_map->partitions[partition].lock, state);
+                continue;
+            }
+
+            // Compute the start of the entry from the cold list entry.
+            ebpf_list_entry_t* list_entry = lru_map->partitions[partition].cold_list.Flink - partition;
+            ebpf_lru_entry_t* entry = (ebpf_lru_entry_t*)list_entry;
+
+            // The effective age of an entry is the maximum of the last used time of all partitions.
+            // Compute the highest timestamp for this entry.
```
Comment (saxena-anurag) [inline libs/execution_context/ebpf_maps.c:1381]: Just a comment: depending on the way / order the entries are accessed on each CPU, this whole loop may run for a very long time before we find an entry to delete. So the worst-case performance may be very bad.

**Comment by Alan-Jowett** [comment]
Disabling ebpf_performance until https://github.com/microsoft/ebpf-for-windows/issues/3486 is fixed.

File: libs/execution_context/ebpf_maps.c (hunk 14 above)
Comment (Alan-Jowett) [reply libs/execution_context/ebpf_maps.c:1206]: Yes. If the entry was never inserted into a partitions hot list it will be marked as unintialized.

File: libs/execution_context/ebpf_maps.c (hunk 15 above)
Comment (Alan-Jowett) [reply libs/execution_context/ebpf_maps.c:1388]: Agreed. I would recommend we keep an eye on this and if there are complaints about LRU replacement time we revisit this.

**Review by mikeagun — DISMISSED** [review]

**Comment by Alan-Jowett** [comment]
Converting to draft until hot list corruption is root caused.

**Comment by Alan-Jowett** [comment]
Reenable user-mode perf tests as they caught a bug in the LRU code.

**Comment by Alan-Jowett** [comment]
Crash has been resolved. Marking as ready for review.

File: libs/execution_context/ebpf_maps.c [hunk 16]
```diff
@@ -1175,6 +1193,7 @@ _initialize_lru_entry(
     ebpf_assert(_get_key_state(map, partition, entry) == EBPF_LRU_KEY_UNINITIALIZED);
```
Comment (mikeagun) [inline libs/execution_context/ebpf_maps.c:1193]: Is this assert needed (since we assert uninitialized for all partitions below)?

Comment (Alan-Jowett) [reply libs/execution_context/ebpf_maps.c:1193]: Nope, it's redundant. Will remove

**Review by mikeagun — APPROVED** [review]

**Review by mikeagun — APPROVED** [review]

File: libs/execution_context/ebpf_maps.c [hunk 17]
```diff
@@ -1004,59 +1098,82 @@ _get_key_state(_In_ const ebpf_core_lru_map_t* map, _In_ const ebpf_lru_entry_t*
...
+        EBPF_LRU_ENTRY_GENERATION_PTR(map, entry)[partition] = map->partitions[partition].current_generation;
+        EBPF_LRU_ENTRY_LAST_USED_TIME_PTR(map, entry)[partition] = ebpf_query_time_since_boot(false);
+        ebpf_list_insert_tail(
+            &map->partitions[partition].hot_list, &EBPF_LRU_ENTRY_LIST_ENTRY_PTR(map, entry)[partition]);
+        map->partitions[partition].hot_list_size++;
+        map->partitions[partition].hot_list_size++;
+        break;
+    case EBPF_LRU_KEY_COLD:
+        // Remove from cold list.
+        EBPF_LRU_ENTRY_GENERATION_PTR(map, entry)[partition] = map->partitions[partition].current_generation;
```
[the remark this answers is not in the document]
Comment (Alan-Jowett) [reply libs/execution_context/ebpf_maps.c:1159]: the [partition] is the deref.

File: libs/execution_context/ebpf_maps.c [hunk 18]
```diff
@@ -1004,59 +1098,82 @@ _get_key_state(_In_ const ebpf_core_lru_map_t* map, _In_ const ebpf_lru_entry_t*
...
+        EBPF_LRU_ENTRY_LAST_USED_TIME_PTR(map, entry)[partition] = ebpf_query_time_since_boot(false);
+        ebpf_list_remove_entry(&EBPF_LRU_ENTRY_LIST_ENTRY_PTR(map, entry)[partition]);
+        ebpf_list_insert_tail(
+            &map->partitions[partition].hot_list, &EBPF_LRU_ENTRY_LIST_ENTRY_PTR(map, entry)[partition]);
+        map->partitions[partition].hot_list_size++;
```
[the remark this answers is not in the document]
Comment (Alan-Jowett) [reply libs/execution_context/ebpf_maps.c:1164]: Good catch. That's a bug.

File: libs/execution_context/ebpf_maps.c [hunk 19]
```diff
@@ -43,33 +43,126 @@ typedef struct _ebpf_core_object_map
...
+} ebpf_lru_partition_t;
+
+/**
+ * @brief The map definition for an LRU map.
+ */
+typedef struct _ebpf_core_lru_map
+{
+    ebpf_core_map_t core_map; //< Core map structure.
+    size_t partition_count;   //< Number of LRU partitions. Limited to a maximum of EBPF_LRU_MAXIMUM_PARTITIONS.
+    uint8_t padding[24];      //< Required to ensure partitions are cache aligned.
```
Comment (mtfriesen) [inline libs/execution_context/ebpf_maps.c:163]: I'm still not sure about this hard-coded 24 byte padding. Why not
`DECLSPEC_CACHEALIGN ebpf_lru_partition_t partitions[1];`
to ensure the partitions field is cache aligned on any architecture?

Comment (mtfriesen) [reply libs/execution_context/ebpf_maps.c:163]: Oh, the `ebpf_lru_partition_t` type is already cache aligned. The padding field is redundant, then?

File: libs/execution_context/ebpf_maps.c [hunk 20]
```diff
@@ -1212,6 +1369,87 @@ _delete_lru_hash_map(_In_ _Post_invalid_ ebpf_core_map_t* map)
...
+ */
+static ebpf_lru_entry_t*
+_reap_lru_cold_lists(ebpf_core_lru_map_t* lru_map)
+{
+    ebpf_lru_entry_t* oldest_entry = NULL;
+    uint64_t oldest_timestamp = UINT64_MAX;
+    bool non_empty_partition_found = true;
+
+    // Scan all partitions until we find a non-empty cold list or we have an oldest entry.
+    while (!oldest_entry && non_empty_partition_found) {
```
Comment (mtfriesen) [inline libs/execution_context/ebpf_maps.c:1386]: Is this loop guaranteed to terminate? If this routine is entered (i.e. the hash table _was_ full) and then some other thread deletes every entry from the map, won't this loop scan forever? Intuitively, it will not find an oldest entry nor will it find a non-empty partition, and from reading the code, I don't see a limit to the loop iterations on an empty LRU.

**Review by mtfriesen — APPROVED** [review]

File: libs/execution_context/ebpf_maps.c [hunk 21]
```diff
@@ -1119,16 +1250,18 @@ _lru_hash_table_notification(
...
         break;
     case EBPF_HASH_TABLE_NOTIFICATION_TYPE_FREE:
         _uninitialize_lru_entry(lru_map, entry);
         break;
-    case EBPF_HASH_TABLE_NOTIFICATION_TYPE_USE:
-        _insert_into_hot_list(lru_map, entry);
-        break;
+    case EBPF_HASH_TABLE_NOTIFICATION_TYPE_USE: {
+        _insert_into_hot_list(lru_map, partition, entry);
+    } break;
```
Comment (dthaler) [inline libs/execution_context/ebpf_maps.c:1264]: The addition of {} looks odd, especially in line 1264.  Can they be reverted?

File: libs/execution_context/ebpf_maps.c [hunk 22]
```diff
@@ -1179,13 +1326,23 @@ _create_lru_hash_map(
...
+        ebpf_list_initialize(&lru_map->partitions[partition].hot_list);
+        ebpf_list_initialize(&lru_map->partitions[partition].cold_list);
+        ebpf_lock_create(&lru_map->partitions[partition].lock);
 
-    lru_map->current_generation = EBPF_LRU_INITIAL_GENERATION;
-    lru_map->hot_list_size = 0;
-    lru_map->hot_list_limit = max(map_definition->max_entries / EBPF_LRU_GENERATION_COUNT, 1);
+        lru_map->partitions[partition].current_generation = EBPF_LRU_INITIAL_GENERATION;
+        lru_map->partitions[partition].hot_list_size = 0;
+        // Assuming that access is randomly distributed, each partition should have the same number of entries in it's
```
Comment (dthaler) [inline libs/execution_context/ebpf_maps.c:1338]:
```suggestion
        // Assuming that access is randomly distributed, each partition should have the same number of entries in its
```

**Review by mikeagun — APPROVED** [review]
